x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpdateDatafeedAction.java

```diff
@@ -50,34 +63,74 @@ protected String executor() {
 
     @Override
     protected void masterOperation(UpdateDatafeedAction.Request request, ClusterState state,
-                                   ActionListener<PutDatafeedAction.Response> listener) {
+                                   ActionListener<PutDatafeedAction.Response> listener) throws Exception {
         final Map<String, String> headers = threadPool.getThreadContext().getHeaders();
 
-        clusterService.submitStateUpdateTask("update-datafeed-" + request.getUpdate().getId(),
-                new AckedClusterStateUpdateTask<PutDatafeedAction.Response>(request, listener) {
-                    private volatile DatafeedConfig updatedDatafeed;
+        // Check datafeed is stopped
+        PersistentTasksCustomMetaData tasks = state.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
+        if (MlTasks.getDatafeedTask(request.getUpdate().getId(), tasks) != null) {
+            listener.onFailure(ExceptionsHelper.conflictStatusException(
+                    Messages.getMessage(Messages.DATAFEED_CANNOT_UPDATE_IN_CURRENT_STATE,
+                            request.getUpdate().getId(), DatafeedState.STARTED)));
+            return;
+        }
+
+        String datafeedId = request.getUpdate().getId();
+
+        CheckedConsumer<Boolean, Exception> updateConsumer = ok -> {
+            datafeedConfigProvider.updateDatefeedConfig(request.getUpdate().getId(), request.getUpdate(), headers,
+                    this::validateDatafeedJob,
+                    ActionListener.wrap(
+                            updatedConfig -> listener.onResponse(new PutDatafeedAction.Response(updatedConfig)),
+                            listener::onFailure
+                    ));
+        };
 
-                    @Override
-                    protected PutDatafeedAction.Response newResponse(boolean acknowledged) {
-                        if (acknowledged) {
-                            logger.info("Updated datafeed [{}]", request.getUpdate().getId());
-                        }
-                        return new PutDatafeedAction.Response(updatedDatafeed);
+
+        if (request.getUpdate().getJobId() != null) {
+            checkJobDoesNotHaveADifferentDatafeed(request.getUpdate().getJobId(), datafeedId,
+                    ActionListener.wrap(updateConsumer, listener::onFailure));
+        } else {
+            updateConsumer.accept(Boolean.TRUE);
+        }
+    }
+
+    private void validateDatafeedJob(DatafeedConfig updatedConfig, ActionListener<Boolean> listener) {
```

Comment:
This method seems like it could be in `JobConfigProvider`? Then we don't have to duplicate it between this action and the put-datafeed action.